internal/service/resourcepolicy/model.go

```diff
@@ -0,0 +1,85 @@
+package resourcepolicy
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"go.mongodb.org/atlas-sdk/v20240805003/admin"
+)
+
+func NewTFResourcePolicyModel(ctx context.Context, input *admin.ApiAtlasResourcePolicy) (*TFResourcePolicyModel, diag.Diagnostics) {
+	createdByUser, diags := NewUserMetadataObjectType(ctx, input.CreatedByUser)
+	if diags.HasError() {
+		return nil, diags
+	}
+	lastUpdatedByUser, diags := NewUserMetadataObjectType(ctx, input.LastUpdatedByUser)
+	if diags.HasError() {
+		return nil, diags
+	}
+	policies, diags := NewPolicyObjectType(ctx, input.Policies)
+	if diags.HasError() {
+		return nil, diags
+	}
+	return &TFResourcePolicyModel{
+		CreatedByUser:     *createdByUser,
+		CreatedDate:       types.StringPointerValue(conversion.TimePtrToStringPtr(input.CreatedDate)),
+		ID:                types.StringPointerValue(input.Id),
+		LastUpdatedByUser: *lastUpdatedByUser,
+		LastUpdatedDate:   types.StringPointerValue(conversion.TimePtrToStringPtr(input.LastUpdatedDate)),
+		Name:              types.StringPointerValue(input.Name),
+		OrgID:             types.StringPointerValue(input.OrgId),
+		Policies:          *policies,
+		Version:           types.StringPointerValue(input.Version),
+	}, nil
+}
+
+func NewUserMetadataObjectType(ctx context.Context, input *admin.ApiAtlasUserMetadata) (*types.Object, diag.Diagnostics) {
+	if input == nil {
```

Comment:
is it ok to call NewUserMetadataObjectType with nil input or it should be considered a logical error? same for NewPolicyObjectType

Reply:
It is ok. Added a test case in ac5d66c7237c050dedbc0c5cd4cb613ee791a19a